crates/bevy_app/src/app.rs

```diff
@@ -1068,4 +1103,60 @@ mod tests {
         }
         App::new().add_plugin(PluginRun);
     }
+
+    mod system_stepping {
+        use super::*;
+        use crate::app::{EventWriter, ScheduleEvent};
+        use crate::CoreSchedule;
+        use bevy_ecs::prelude::*;
+        use bevy_ecs::schedule::ExecutorKind;
+
+        #[derive(Resource, Default)]
+        struct ScheduleEvents(Vec<ScheduleEvent>);
+
+        fn send_schedule_event(
+            mut events: ResMut<ScheduleEvents>,
+            mut writer: EventWriter<ScheduleEvent>,
+        ) {
+            for event in events.0.drain(..) {
+                writer.send(event);
+            }
+        }
+
+        fn push_event(app: &mut App, event: ScheduleEvent) {
+            let mut events = app
+                .world
+                .get_resource_mut::<ScheduleEvents>()
+                .expect("ScheduleEvents resource not found in world");
+            events.0.push(event);
+        }
+
+        fn get_stepping(app: &App) -> bool {
```

Comment:
This feels like there should be a `SteppingSettings` enum that's used as a field on both `Schedule` and `App`. Then either make that `pub` if writing it is always safe, or add a `&self` getter method.

Reply:
From the idea of `SteppingSettings` on `App`, we can do a "global stepping enabled" which applies to all `Schedule`s, but we cannot globally step a single system in that case.  The problem is that stepping a single system requires us to limit the step to a specific `Schedule`, otherwise more than one system will run.  To do a single system in single `Schedule` step, we'd need to specify the schedule label in the `App`s `SteppingSettings`.

As a side note, global single-frame stepping would work just fine.

Multiple schedules just make things very complicated from the user perspective.  I really was lucky in that the breakout example used a different schedule from the core input & render systems.

Ideally, for the user API, it would be wonderful to be able to just send a generic `StepSystem` event, with no schedule label attached, and have it only be sent to one `Schedule`, but I don't know how to do that right now.  Selecting the "next" `Schedule` that should receive the step seems complex at the moment.

I'd like to get the API to that place, but it should probably be a follow-on PR.

Reply:
It feels like we should be able to store the `ScheduleSettings` on each `Schedule` TBH? And then we can just modify what `Schedule::run` does. I'm actually not sure that this should be stored on `App` at all now.

Reply:
Ok, so I got the `StepState` moved into `SystemSchedule`.  All the methods for manipulating it are on `Schedule`.  I did it this way because `SystemSchedule` is passed to the `Executor`, and all the existing methods for modifying the schedule do reside in `Schedule`.

Does this kind of match what you were looking for?